[SPARK-27207][SQL] : Ensure aggregate buffers are initialized again for So…

pgandhi999 opened the pull request:
…rtBasedAggregate

Normally, the aggregate operations that are invoked for an aggregation buffer for User Defined Aggregate Functions(UDAF) follow the order like initialize(), update(), eval() OR initialize(), merge(), eval(). However, after a certain threshold configurable by spark.sql.objectHashAggregate.sortBased.fallbackThreshold is reached, ObjectHashAggregate falls back to SortBasedAggregator which invokes the merge or update operation without calling initialize() on the aggregate buffer.

## What changes were proposed in this pull request?

The fix here is to initialize aggregate buffers again when fallback to SortBasedAggregate operator happens.

## How was this patch tested?

The patch was tested as part of [SPARK-24935](https://issues.apache.org/jira/browse/SPARK-24935) as documented in PR https://github.com/apache/spark/pull/23778.

Commit:
[SPARK-27207] : Ensure aggregate buffers are initialized again for SortBasedAggregate

Normally, the aggregate operations that are invoked for an aggregation buffer for User Defined Aggregate Functions(UDAF) follow the order like initialize(), update(), eval() OR initialize(), merge(), eval(). However, after a certain threshold configurable by spark.sql.objectHashAggregate.sortBased.fallbackThreshold is reached, ObjectHashAggregate falls back to SortBasedAggregator which invokes the merge or update operation without calling initialize() on the aggregate buffer.

The fix here is to initialize aggregate buffers again when fallback to SortBasedAggregate operator happens.

pgandhi999 commented:
cc @cloud-fan This is a bug with SortBasedAggregate that was exposed while testing PR #23778 . Have filed a separate JIRA alongwith the PR here. Request you to review it. Thank you.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

pgandhi999 commented:
ok to test

SparkQA commented:
**[Test build #103691 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103691/testReport)** for PR 24149 at commit [`400db3d`](https://github.com/apache/spark/commit/400db3d1dad4b92e0a70b89c2c687a42017d589b).

SparkQA commented:
**[Test build #103691 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103691/testReport)** for PR 24149 at commit [`400db3d`](https://github.com/apache/spark/commit/400db3d1dad4b92e0a70b89c2c687a42017d589b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103691/
Test PASSed.

cloud-fan commented:
I'm a little confused, we reuse the UDAFs when falling back to sort based agg, why do we need to initialize the UDAF again?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala:
```diff
@@ -59,17 +59,21 @@ class ObjectAggregationIterator(
 
   // Hacking the aggregation mode to call AggregateFunction.merge to merge two aggregation buffers
-  private val mergeAggregationBuffers: (InternalRow, InternalRow) => Unit = {
+  var (sortBasedAggExpressions, sortBasedAggFunctions): (
```
why it's `var` instead of `val`?

pgandhi999 replied:
Changed it to val

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala:
```diff
@@ -106,26 +110,28 @@ class ObjectAggregationIterator(
-  private def createNewAggregationBuffer(): SpecificInternalRow = {
-    val bufferFieldTypes = aggregateFunctions.flatMap(_.aggBufferAttributes.map(_.dataType))
+  private def createNewAggregationBuffer(
+    functions: Array[AggregateFunction]): SpecificInternalRow = {
```
4 space indentation here.

pgandhi999 replied:
Done

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala:
```diff
@@ -106,26 +110,28 @@ class ObjectAggregationIterator(
   }
 
-  private def initAggregationBuffer(buffer: SpecificInternalRow): Unit = {
+  private def initAggregationBuffer(
```
it's only called once, let's inline it

pgandhi999 replied:
Done

cloud-fan commented:
good catch! Can we add a UT? we can create a `TypedImperativeAggregate` implementation which fails if initialize is not called.

Commit:
[SPARK-27207] : Adding Unit Test and addressing reviews

Adding unit test and refactoring code

pgandhi999 commented:
@cloud-fan Thank you for reviewing. Have added a unit test which was failing before the code change.

SparkQA commented:
**[Test build #103773 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103773/testReport)** for PR 24149 at commit [`ea050f7`](https://github.com/apache/spark/commit/ea050f7547597f82e591cc98085feefe58c4ed79).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103773/
Test FAILed.

Commit:
[SPARK-27207] : Fixing Scalastyle tests

SparkQA commented:
**[Test build #103775 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103775/testReport)** for PR 24149 at commit [`0714876`](https://github.com/apache/spark/commit/07148763e5f8aa83f846ccdc165fd1fb4907f2da).

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/TypedImperativeAggregateSuite.scala:
```diff
@@ -299,5 +319,86 @@ object TypedImperativeAggregateSuite {
+    var maxValueBuffer: MaxValue = null
+    override def createAggregationBuffer(): MaxValue = {
+      // Returns Int.MinValue if all inputs are null
+      maxValueBuffer = new MaxValue(Int.MinValue)
```
why do we need to save it to a member variable? I think the bug can be exposed even if we just return the buffer here.

pgandhi999 replied:
@cloud-fan I am still looking more into it, but for some reason, calling merge() without invoking initialize() does not cause any visible exception on normal UDAF functions, but it fails with a Null Pointer Exception when I test it with the test case described in SPARK-24935(PR #24144 ). My guess is that for the above test case, since, two different aggregation buffer instances are created(SketchState and UnionState), the exception shows up. Will investigate more on it and get back to you soon. Thank you.

shaneknapp commented:
test this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/9156/
Test PASSed.

SparkQA commented:
**[Test build #103782 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103782/testReport)** for PR 24149 at commit [`0714876`](https://github.com/apache/spark/commit/07148763e5f8aa83f846ccdc165fd1fb4907f2da).

SparkQA commented:
**[Test build #103782 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/103782/testReport)** for PR 24149 at commit [`0714876`](https://github.com/apache/spark/commit/07148763e5f8aa83f846ccdc165fd1fb4907f2da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/103782/
Test PASSed.

Commit:
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-27207

[SPARK-27207] : Upmerging with master branch

pgandhi999 commented:
@cloud-fan Regarding our discussion in PR #24144 , I just found out a case where Spark initializes a UDAF, runs `update` and then runs `merge`. It happens in `SortBasedAggregator`. So, the code blows up in this case. The code in `ObjectAggregationIterator.scala` is pasted below:

```
// Two-way merges initialAggBufferIterator and inputIterator
      private def findNextSortedGroup(): Boolean = {
        if (hasNextInput || hasNextAggBuffer) {
          // Find smaller key of the initialAggBufferIterator and initialAggBufferIterator
          groupingKey = findGroupingKey()
          result = new AggregationBufferEntry(groupingKey, makeEmptyAggregationBuffer)

          // Firstly, update the aggregation buffer with input rows.
          while (hasNextInput &&
            groupingKeyOrdering.compare(inputIterator.getKey, groupingKey) == 0) {
            processRow(result.aggregationBuffer, inputIterator.getValue)
            hasNextInput = inputIterator.next()
          }

          // Secondly, merge the aggregation buffer with existing aggregation buffers.
          // NOTE: the ordering of these two while-block matter, mergeAggregationBuffer() should
          // be called after calling processRow.
          while (hasNextAggBuffer &&
            groupingKeyOrdering.compare(initialAggBufferIterator.getKey, groupingKey) == 0) {
            mergeAggregationBuffers(result.aggregationBuffer, initialAggBufferIterator.getValue)
            hasNextAggBuffer = initialAggBufferIterator.next()
          }

          true
        } else {
          false
        }
      }
```

It calls `update` first and then calls `merge` on the same buffer. I found out the issue while testing this PR today.

cloud-fan commented:
But we call `update` and `merge` for different copies of the aggregate functions, don't we?

pgandhi999 commented:
Turns out that is not true in the case of `SortBasedAggregator`. In the code that I have pasted above, `processRow()` is called on `result.aggregationBuffer` which performs `update` and later, `mergeAggregationBuffers` is called on the same `result.aggregationBuffer` which performs `merge`. This is what I could infer from the code as well as debug logs that I added, correct me if I am wrong.

cloud-fan commented:
Yea it's the same buffer instance, but not same aggregate function instance, IIUC.

pgandhi999 commented:
@cloud-fan So after going through the code, I see that we are calling update and merge for different copies of aggregatefunctions but are using the buffer created for one copy of aggregatefunction. I am really not an expert with the aggregation framework so was wondering if you could guide me here by elaborating more about how Spark aggregate functions use the aggregation buffer? Thank you.

cloud-fan commented:
Each aggregate function will create its own buffer. When we feed a buffer to a agg func, we are not asking the agg func to replace its own buffer with the new one, but we ask it to merge the new buffer to its own buffer.

Commit:
[SPARK-27207] : Fix SortBasedAggregator to run with different aggregate functions and write unit test

Fix SortBasedAggregator to ensure that update and merge are performed with two different sets of aggregate functions, one for update and one for merge respectively.

Commit:
[SPARK-27207] : Fix new line for TypedImperativeAggregateSuite

SparkQA commented:
**[Test build #104090 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104090/testReport)** for PR 24149 at commit [`088cbc6`](https://github.com/apache/spark/commit/088cbc692822a545a5063c482defed0f5491fcfc).

pgandhi999 commented:
@cloud-fan So I realized that the bug was caused as I was creating the aggregate buffer for `sortBasedMergeAggFunctions`, calling `processRow` on `update` operation with `aggregateFunctions` and then calling `merge` once again with `sortBasedMergeAggFunctions`. Fixed the bug by having a separate update buffer initialized with `aggregateFunctions`, perform the `update` on that buffer, merge it's results to a new aggregate buffer initialized for `sortBasedMergeAggFunctions`, finally on which `merge` is called and the final result is returned. It may not be the best solution so your valuable guidance in this matter is really appreciated. Thank you.

SparkQA commented:
**[Test build #104090 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104090/testReport)** for PR 24149 at commit [`088cbc6`](https://github.com/apache/spark/commit/088cbc692822a545a5063c482defed0f5491fcfc).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104090/
Test FAILed.

Commit:
[SPARK-27207] : Changing design to use one buffer but initializing for different aggregate functions

SparkQA commented:
**[Test build #104099 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104099/testReport)** for PR 24149 at commit [`6a5ed71`](https://github.com/apache/spark/commit/6a5ed71081060688fabef19ceae39b597980dfc2).

SparkQA commented:
**[Test build #104099 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104099/testReport)** for PR 24149 at commit [`6a5ed71`](https://github.com/apache/spark/commit/6a5ed71081060688fabef19ceae39b597980dfc2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104099/
Test FAILed.

Commit:
Revert "[SPARK-27207] : Changing design to use one buffer but initializing for different aggregate functions"

This reverts commit 6a5ed71081060688fabef19ceae39b597980dfc2.

Reverting to previous commit

Commit:
[SPARK-27207] : Revert to previous commit and serialize the buffer with sortBasedMergeAggregateFunctions

Use two aggregate buffers as in previous commit and serialize updateResult on sortBasedMergeAggFunctions before merging it with finalResult.

SparkQA commented:
**[Test build #104326 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104326/testReport)** for PR 24149 at commit [`433b1bb`](https://github.com/apache/spark/commit/433b1bb5e1a4a53f604bd93868d341e8d14ceb62).

SparkQA commented:
**[Test build #104326 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104326/testReport)** for PR 24149 at commit [`433b1bb`](https://github.com/apache/spark/commit/433b1bb5e1a4a53f604bd93868d341e8d14ceb62).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104326/
Test FAILed.

pgandhi999 commented:
test this please

SparkQA commented:
**[Test build #104335 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104335/testReport)** for PR 24149 at commit [`433b1bb`](https://github.com/apache/spark/commit/433b1bb5e1a4a53f604bd93868d341e8d14ceb62).

SparkQA commented:
**[Test build #104335 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104335/testReport)** for PR 24149 at commit [`433b1bb`](https://github.com/apache/spark/commit/433b1bb5e1a4a53f604bd93868d341e8d14ceb62).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104335/
Test FAILed.

pgandhi999 commented:
@shaneknapp I could be wrong here but it does not seem like the failing pyspark tests have anything to do with my PR. WDYT?

Commit:
[SPARK-27207] : Upmerging with master branch
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-27207

SparkQA commented:
**[Test build #104448 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104448/testReport)** for PR 24149 at commit [`446e182`](https://github.com/apache/spark/commit/446e182a79ab789d6ce28afc9a6b89655b7adaa4).

SparkQA commented:
**[Test build #104448 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/104448/testReport)** for PR 24149 at commit [`446e182`](https://github.com/apache/spark/commit/446e182a79ab789d6ce28afc9a6b89655b7adaa4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/104448/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala:
```diff
@@ -258,21 +275,29 @@ class SortBasedAggregator(
           // NOTE: the ordering of these two while-block matter, mergeAggregationBuffer() should
           // be called after calling processRow.
           while (hasNextAggBuffer &&
             groupingKeyOrdering.compare(initialAggBufferIterator.getKey, groupingKey) == 0) {
```
I think the data flow is
1. start with hash agg, create and initialize an empty buffer for each group key, and consume input records by calling `update`
2. fallback to sort based agg
3. create and initialize an empty buffer for each group key, and consume input records by calling `update`.
4. when all the input records for one group key are consumed up, call `merge` to merge the agg buffer that is generated by hash agg.

This data flow is fine most of the time, but is problematic for hive UDAF, because it doesn't support INIT -> UPDATE -> MERGE -> FINISH.

That said, this patch may cause perf regression for normal UDAFs. We should only do it for hive UDAF, to switch from `INIT -> UPDATE -> MERGE -> FINISH` to `INIT -> UPDATE -> FINISH, INIT -> MERGE -> FINISH`

m44444 replied:
Hi @cloud-fan, what you said here is really killing! With Spark 2.4.1 by turning this conf off I see the DataSketches hll issue is solved: `--conf spark.sql.execution.useObjectHashAggregateExec=false`. Which basically disable the hash agg attempt. But as you said, this is downgrading the general performance (not so bad when people have a sense about how big their data is).
However my question is, is it possible that it can be automatically done for the framework to recognize Hive UDAF and only apply its way to it? e.g. For such a query `select year, count(month), hive_udaf_count(month) from ... group by year` I want the Spark count() to behave as the way INIT -> UPDATE -> MERGE -> FINISH, while the hive_udaf_count() to behave as the way INIT -> UPDATE -> FINISH, INIT -> MERGE -> FINISH

pgandhi999 commented:
Thank you for your comments. I am involved in Spark Summit this week. Will get back to the PR next week. @cloud-fan wondering if you are here as well. Thank you.

cloud-fan commented:
Hi @pgandhi999 unfortunately I can't this summit for some personal reasons, hope you enjoy it!

BTW I came up with a new idea to fix this problem in https://github.com/apache/spark/pull/24459

Commit:
[SPARK-27207] : Reverting the two buffer logic and simplifying the code

Since, https://github.com/apache/spark/pull/24459 fixes the init-update-merge issue, the fix here is reverted.

Commit:
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-27207

[SPARK-27207] : Upmerging with master branch

SparkQA commented:
**[Test build #105032 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105032/testReport)** for PR 24149 at commit [`db46cf7`](https://github.com/apache/spark/commit/db46cf702f18599bd302d4b3842eb805c8490a42).

SparkQA commented:
**[Test build #105032 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105032/testReport)** for PR 24149 at commit [`db46cf7`](https://github.com/apache/spark/commit/db46cf702f18599bd302d4b3842eb805c8490a42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105032/
Test PASSed.

pgandhi999 commented:
@cloud-fan Have updated the PR. Thank you.

pgandhi999 commented:
Hello @cloud-fan , WDYT about the updated PR? Thank you.

cloud-fan commented on sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveUDAFSuite.scala:
```diff
@@ -123,6 +133,26 @@ class HiveUDAFSuite extends QueryTest with TestHiveSingleton with SQLTestUtils {
     }
   }
 
+  test("SPARK-27207: customized Hive UDAF with two aggregation buffers for Sort" +
```
does this test fail without your patch? I think we can write a test with a custom UDAF which fails without initialization, but Hive UDAF does not fail without initialization.

pgandhi999 replied:
So it used to fail earlier without my patch, but your latest patch seems to have fixed it. Will come up with another test case. Thank you.

Commit:
[SPARK-27207] : Coming up with a unit test for custom UDAF

SparkQA commented:
**[Test build #105167 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105167/testReport)** for PR 24149 at commit [`df330fa`](https://github.com/apache/spark/commit/df330fa1cf45e4533d93a8aebade12a504b6726b).

SparkQA commented:
**[Test build #105167 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105167/testReport)** for PR 24149 at commit [`df330fa`](https://github.com/apache/spark/commit/df330fa1cf45e4533d93a8aebade12a504b6726b).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105167/
Test FAILed.

Commit:
[SPARK-27207] : Fixing Scalastyle Tests

SparkQA commented:
**[Test build #105168 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105168/testReport)** for PR 24149 at commit [`5bd474c`](https://github.com/apache/spark/commit/5bd474cd55f6f77fbd003c532bd77343e42750bf).

SparkQA commented:
**[Test build #105168 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105168/testReport)** for PR 24149 at commit [`5bd474c`](https://github.com/apache/spark/commit/5bd474cd55f6f77fbd003c532bd77343e42750bf).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105168/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/TypedImperativeAggregateSuite.scala:
```diff
@@ -299,5 +319,87 @@ object TypedImperativeAggregateSuite {
+   * in aggregation buffer.
+   */
+  private case class TypedMax2(
+    child: Expression,
```
4 space indentation.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/TypedImperativeAggregateSuite.scala:
```diff
@@ -299,5 +319,87 @@ object TypedImperativeAggregateSuite {
+   * Calculate the max value with object aggregation buffer. This stores class MaxValue
+   * in aggregation buffer.
+   */
+  private case class TypedMax2(
```
can we simplify it? I think we just need to do some initialization work in  `createAggregationBuffer`.
```
case class MyUDAF ... {
  var initialized = false
  override def createAggregationBuffer(): MyBuffer = {
    initialized = true
    null
  }

  override def update(buffer: MaxValue, input: InternalRow): MyBuffer = {
    assert(initialized)
    null
  }

  ...
}
```

Commit:
[SPARK-27207] : Simplifying unit test and indentation

SparkQA commented:
**[Test build #105221 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105221/testReport)** for PR 24149 at commit [`006616e`](https://github.com/apache/spark/commit/006616ed3c82ef2f7cac1631d9306e34fd069cff).

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/TypedImperativeAggregateSuite.scala:
```diff
@@ -210,6 +211,14 @@ class TypedImperativeAggregateSuite extends QueryTest with SharedSQLContext {
+    withSQLConf("spark.sql.objectHashAggregate.sortBased.fallbackThreshold" -> "5") {
+      val df = data.toDF("value", "key").coalesce(2)
+      val query = df.groupBy($"key").agg(typedMax2($"value"), count($"value"), typedMax2($"value"))
+      query.show(10, false)
```
nit: we should not use `show` in the test. Let's use `checkAnswer`.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/TypedImperativeAggregateSuite.scala:
```diff
@@ -299,5 +313,62 @@ object TypedImperativeAggregateSuite {
+      null
+    }
+
+    override lazy val deterministic: Boolean = true
```
is this needed?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/ObjectAggregationIterator.scala:
```diff
@@ -58,18 +58,23 @@ class ObjectAggregationIterator(
+    (newExpressions, initializeAggregateFunctions(newExpressions, 0))
+  }
+
+  // Hacking the aggregation mode to call AggregateFunction.merge to merge two aggregation buffers
```
I think this comment should be put before
```
val newExpressions = aggregateExpressions.map {
      case agg @ AggregateExpression(_, Partial, _, _) =>
        agg.copy(mode = PartialMerge)
      case agg @ AggregateExpression(_, Complete, _, _) =>
        agg.copy(mode = Final)
      case other => other
    }
```

cloud-fan commented:
After some more thoughts, I have some different ideas now.

I checked the `TypedImperativeAggregate` implementations, the `initialize` method is used to initialize the aggregate buffer, not to initialize themselves. That said, `TypedImperativeAggregate` implementations should be stateless.

Come back to this bug, it can only be exposed if the UDAF needs to initialize itself, which should not be allowed. I think we can just add some doc to `TypedImperativeAggregate`, saying that it must be stateless. Sorry for the back and forth!

pgandhi999 commented:
@cloud-fan I do see your point but don't you still think, if fallback to SortBasedAggregate occurs, the initialize method for the new aggregate function should be invoked as I recall you said earlier:

> Each aggregate function will create its own buffer. When we feed a buffer to a agg func, we are not asking the agg func to replace its own buffer with the new one, but we ask it to merge the new buffer to its own buffer.

By the above logic, we still need to ensure that the respective aggregate buffer is initialized for the new set of aggregate functions. Or we should just use the same set of aggregate functions again for SortBasedAggregate. Correct me if I am wrong here.

cloud-fan commented:
Each aggregate function will create its own buffer, but the aggregate function doesn't hold the buffer, the buffer is managed by Spark. Aggregate function should be stateless.

Let's say we have an aggregate expression `expr`. `expr` creates an aggregate function `f1` to do the work before sort fallback happens. `f1` creates a buffer and starts to accumulate the buffer. When sort fallback happens, `expr` creates a new aggregate function `f2`. We still ask `f1` to create the buffer, and then `f2` starts working and accumulate the buffer.

Since `f1` and `f2` are the same functions, this should be fine.

SparkQA commented:
**[Test build #105221 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105221/testReport)** for PR 24149 at commit [`006616e`](https://github.com/apache/spark/commit/006616ed3c82ef2f7cac1631d9306e34fd069cff).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105221/
Test PASSed.

pgandhi999 commented:
I see. Sure I can go ahead and close this PR. Thank you.

cloud-fan commented:
@pgandhi999 it could be great if you can update the classdoc of `AggregateFunction` and say that it should be stateless. thanks!

Commit:
[SPARK-27207] : Reverting changes and updating doc

Commit:
Merge branch 'master' of https://github.com/pgandhi999/spark into SPARK-27207

[SPARK-27207] : Upmerging with master branch

pgandhi999 commented:
@cloud-fan I have updated the doc. If it also needs to be updated someplace else, do let me know. Thank you.

SparkQA commented:
**[Test build #105257 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105257/testReport)** for PR 24149 at commit [`28ea0f9`](https://github.com/apache/spark/commit/28ea0f9e9fefe7a23ab43843286812a68e9fa7b9).

SparkQA commented:
**[Test build #105257 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105257/testReport)** for PR 24149 at commit [`28ea0f9`](https://github.com/apache/spark/commit/28ea0f9e9fefe7a23ab43843286812a68e9fa7b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105257/
Test FAILed.

pgandhi999 commented:
test this please.

SparkQA commented:
**[Test build #105265 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105265/testReport)** for PR 24149 at commit [`28ea0f9`](https://github.com/apache/spark/commit/28ea0f9e9fefe7a23ab43843286812a68e9fa7b9).

SparkQA commented:
**[Test build #105265 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/105265/testReport)** for PR 24149 at commit [`28ea0f9`](https://github.com/apache/spark/commit/28ea0f9e9fefe7a23ab43843286812a68e9fa7b9).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/105265/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
